inst/include/plant/patch.h

```diff
@@ -295,6 +337,57 @@ ode::const_iterator Patch<T,E>::set_ode_state(ode::const_iterator it,
   return it;
 }
 
+// pre-cached environments, used for mutant runs
+template <typename T, typename E>
+ode::const_iterator Patch<T,E>::set_ode_state(ode::const_iterator it,
+                                              int index) {
+
+  it = ode::set_ode_state(species.begin(), species.end(), it);
+
+  environment = environment_cache[index];
```

Comment:
Here and below on Lin 377, I suspect we're wasting a lot of computing time copying environment objects. When you call `environment = environment_cache[index]` it will copy the stuff in the environment_cache into the `environment` object. 

We could speed things up a lot by making `environment` a pointer rather than an object itself. In that case you you wouldn't copy it, you'd just point the pointer at the right spot in the cache. Pointers are tiny and cheap, this would avoid copying, so is potentially much faster.

It would look something like 

```
*environment = &environment_cache[index];
```

To make this work we need to change entire environment object to a pointer. Then allocate appropriate memory when running a resident.

Reply:
We'd also need to change the way env is passed into functions. 

It's definitely the right thing to do, as copying env objects seems really wasteful and slow